src/bvh/bvh.rs

```diff
@@ -0,0 +1,401 @@
+            }
+        }
+
+        // Create six buckets, and six index assignment vectors
+        const NUM_BUCKETS: usize = 6;
+        let mut buckets = [Bucket::empty(); NUM_BUCKETS];
+        let mut bucket_assignments: [Vec<usize>; NUM_BUCKETS] = Default::default();
+
+        if split_axis_size < EPSILON {
+            // Spread the remaining shapes evenly across buckets,
+            // instead of using a heuristic
+            for (counter, idx) in (&indices).iter().enumerate() {
+                let shape = &shapes[*idx];
+                let shape_aabb = shape.aabb();
+
+                let bucket_num = counter % NUM_BUCKETS;
+
+                buckets[bucket_num].add_aabb(&shape_aabb);
+                bucket_assignments[bucket_num].push(*idx);
+            }
+        } else {
+            // Assign each shape to a bucket
+            for idx in &indices {
+                let shape = &shapes[*idx];
+                let shape_aabb = shape.aabb();
+                let shape_center = shape_aabb.center();
+
+                // Get the relative position of the shape centroid [0.0..1.0]
+                let bucket_num_relative =
+                    (shape_center[split_axis] - centroid_bounds.min[split_axis]) / split_axis_size;
+
+                // Convert that to the actual `Bucket` number
+                let bucket_num = (bucket_num_relative * (NUM_BUCKETS as f32 - 0.01)) as usize;
+
+                // Extend the selected `Bucket` and add the index to the actual bucket
+                buckets[bucket_num].add_aabb(&shape_aabb);
+                bucket_assignments[bucket_num].push(*idx);
+            }
+        }
+
+        // Compute the costs for each configuration and
+        // select the configuration with the minimal costs
+        let mut min_bucket = 0;
+        let mut min_cost = f32::INFINITY;
+        let mut child_l_aabb = AABB::empty();
+        let mut child_r_aabb = AABB::empty();
+        for i in 0..(NUM_BUCKETS - 1) {
+            let (l_buckets, r_buckets) = buckets.split_at(i + 1);
+            let child_l = l_buckets.iter().fold(Bucket::empty(), join_bucket);
+            let child_r = r_buckets.iter().fold(Bucket::empty(), join_bucket);
+
+            let cost = (child_l.size as f32 * child_l.aabb.surface_area() +
+                        child_r.size as f32 * child_r.aabb.surface_area()) /
+                       aabb_bounds.surface_area();
+
+            if cost < min_cost {
+                min_bucket = i;
+                min_cost = cost;
+                child_l_aabb = child_l.aabb;
+                child_r_aabb = child_r.aabb;
+            }
+        }
+
+        fn join_vectors(vectors: &mut [Vec<usize>]) -> Vec<usize> {
+            let mut result = Vec::new();
+            for mut vector in vectors.iter_mut() {
+                result.append(&mut vector);
+            }
+            result
+        }
+
+        // Join together all index buckets, and proceed recursively
```

Comment:
A bucket (in SAH) defines a partition containing both the metadata, such as the union aabb, the cost, a.s.o. AND the shapes. We have split this into two containers. the one with the metadata and one with the indices.